unix/modjni.c

```diff
@@ -0,0 +1,375 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <assert.h>
+#include <string.h>
+#include <errno.h>
+#include <dlfcn.h>
+#include <ffi.h>
+
+#include "py/nlr.h"
+#include "py/runtime.h"
+#include "py/binary.h"
+
+#include <jni.h>
+
+#define JJ(call, ...) (*env)->call(env, __VA_ARGS__)
+
+static JavaVM *jvm;
+static JNIEnv *env;
+static jclass String_class;
+static jmethodID Class_getField_mid;
+static jmethodID Class_getMethods_mid;
+static jmethodID Class_getConstructors_mid;
+static jmethodID Method_getName_mid;
+static jmethodID Method_toString_mid;
+
+STATIC const mp_obj_type_t jobject_type;
+STATIC const mp_obj_type_t jmethod_type;
+
+STATIC mp_obj_t call_method(jobject obj, const char *name, jarray methods, bool is_constr, mp_uint_t n_args, const mp_obj_t *args);
+
+typedef struct _mp_obj_jclass_t {
+    mp_obj_base_t base;
+    jclass cls;
+} mp_obj_jclass_t;
+
+typedef struct _mp_obj_jobject_t {
+    mp_obj_base_t base;
+    jobject obj;
+} mp_obj_jobject_t;
+
+typedef struct _mp_obj_jmethod_t {
+    mp_obj_base_t base;
+    jobject obj;
+    jmethodID meth;
+    qstr name;
+} mp_obj_jmethod_t;
+
+// jclass
+
+STATIC void jclass_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    (void)kind;
+    mp_obj_jclass_t *self = self_in;
+    // Variable value printed as cast to int
+    mp_printf(print, "<jclass @%p>", self->cls);
+}
+
+STATIC void jclass_attr(mp_obj_t self_in, qstr attr_in, mp_obj_t *dest) {
+    if (dest[0] == MP_OBJ_NULL) {
+        // load attribute
+        mp_obj_jclass_t *self = self_in;
+        const char *attr = qstr_str(attr_in);
+
+        jstring field_name = JJ(NewStringUTF, attr);
+        jobject field = JJ(CallObjectMethod, self->cls, Class_getField_mid, field_name);
+        jfieldID field_id = JJ(FromReflectedField, field);
+        jobject obj = JJ(GetStaticObjectField, self->cls, field_id);
+
+        mp_obj_jobject_t *o = m_new_obj(mp_obj_jobject_t);
+        o->base.type = &jobject_type;
+        o->obj = obj;
+        dest[0] = o;
+    }
+}
+
+STATIC mp_obj_t jclass_call(mp_obj_t self_in, mp_uint_t n_args, mp_uint_t n_kw, const mp_obj_t *args) {
+    if (n_kw != 0) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_TypeError, "kwargs not supported"));
```

Comment:
I did not, but doing that now, 1st one takes a lot of args which needs to be prepared, and 2nd is intended for NotImplementedError (where uPy function doesn't take kwargs, while CPy does), so none seems to be too good fit here.